File: jcommons/src/main/java/org/mbari/jcommons/util/Logging.java
/* 
Copyright Monterey Bay Aquarium Research Institute (MBARI) 2022

MBARI licenses this file to you under the Apache License, 
Version 2.0 (the "License"); you may not use this file except in
compliance with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.   
*/
package org.mbari.jcommons.util;

import java.lang.System.Logger;
import java.lang.System.Logger.Level;
import java.util.function.Supplier;

public record Logging(Logger logger, Level level, Throwable cause) {

    public Logging(Logger logger) {
        this(logger, Level.OFF, null);
    }

    public Logging(String loggerName) {
        this(System.getLogger(loggerName), Level.OFF, null);
    }

    public Logging(Class<?> clazz) {
        this(clazz.getName());
    }

    public Logging atTrace() {
        return new Logging(logger, Level.TRACE, cause);
    }

    public Logging atDebug() {
        return new Logging(logger, Level.DEBUG, cause);
    }

    public Logging atInfo() {
        return new Logging(logger, Level.INFO, cause);
    }

    public Logging atWarn() {
        return new Logging(logger, Level.WARNING, cause);
    }

    public Logging atError() {
        return new Logging(logger, Level.ERROR, cause);
    }

    public Logging withCause(Throwable t) {
        return new Logging(logger, level, t);
    }

    public void log(String msg) {
        if (logger.isLoggable(level)) {
            if (cause == null) {
                logger.log(level, msg);
            } else {
                logger.log(level, msg, cause);
            }
        }
    }

    public void log(Supplier<String> fn) {
        if (logger.isLoggable(level)) {
            if (cause == null) {
                logger.log(level, fn);
            } else {
                logger.log(level, fn, cause);
            }
        }
    }
}

File: scommons/src/main/scala/org/mbari/scommons/etc/jdk/Logging.scala
package org.mbari.scommons.etc.jdk

import java.lang.System.Logger
import java.lang.System.Logger.Level
import java.util.function.Supplier

/**
 * Add fluent logging to System.Logger. Usage:
 * {{{
 * import org.fathomnet.support.etc.jdk.Logging.{given, *}
 * given log: Logger = Sytem.getLogger("my.logger")
 *
 * log.atInfo.log("Hello World")
 * log.atInfo.withCause(new RuntimeException("Oops")).log("Hello World")
 *
 * 3.tapLog.atInfo.log(i => "Hello World " + i)
 * }}}
 * * @author Brian Schlining
 */
object Logging:

  case class LoggerBuilder(
      logger: Logger,
      level: Level = Level.OFF,
      throwable: Option[Throwable] = None
  ):

    def atTrace: LoggerBuilder = copy(level = Level.TRACE)
    def atDebug: LoggerBuilder = copy(level = Level.DEBUG)
    def atInfo: LoggerBuilder  = copy(level = Level.INFO)
    def atWarn: LoggerBuilder  = copy(level = Level.WARNING)
    def atError: LoggerBuilder = copy(level = Level.ERROR)

    def withCause(cause: Throwable): LoggerBuilder = copy(throwable = Some(cause))

    def log(msg: String): Unit =
      if (logger.isLoggable(level))
        throwable match
          case Some(e) => logger.log(level, msg, e)
          case None    => logger.log(level, msg)

    def log(fn: Supplier[String]): Unit =
      if (logger.isLoggable(level))
        throwable match
          case Some(e) => logger.log(level, fn, e)
          case None    => logger.log(level, fn)

  given Conversion[Logger, LoggerBuilder] with
    def apply(logger: Logger): LoggerBuilder = LoggerBuilder(logger)

  def apply(name: String): LoggerBuilder    = LoggerBuilder(System.getLogger(name))
    def apply(clazz: Class[?]): LoggerBuilder = apply(clazz.getName)
